capstone/backend/src/test/java/com/google/servlets/GetStoresWithItemTypesServletTest.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2020 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.servlets;
+
+import com.google.spanner.LibraryFunctions;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.List;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletResponse;
+import junit.framework.TestCase;
+import org.junit.Before;
+import org.mockito.Mockito;
+
+public class GetStoresWithItemTypesServletTest extends TestCase {
+  private static final String ITEM_TYPES_KEY = "item_types";
+
+  public void testGetStores() throws ServletException, IOException {
+    String[] ITEM_TYPES = {"MILK"};
+    HashMap<String, String[]> map = new HashMap<>();
+    map.put(ITEM_TYPES_KEY, ITEM_TYPES);
+    SetupObj setupObj = ServletTestUtil.setupMockDataGetList(map);
+
+    GetStoresWithItemTypesServlet servlet = Mockito.spy(GetStoresWithItemTypesServlet.class);
+
+    List<Store> fakeStores = new ArrayList<Store>();
+    Store store1 = new Store(1, "Walmart", "3255 Mission College Blvd");
+    Store store2 = new Store(2, "Target", "4080 Stevens Creek Blvd");
+    Store store3 = new Store(3, "Whole Foods", "301 Ranch Dr");
+    store1.addItem(1, 11.98, "Horizon Organic Whole Shelf-Stable Milk, 8 Oz., 12 Count");
+    store2.addItem(1, 10.38, "Horizon Organic Whole Shelf-Stable Milk, 8 Oz., 12 Count");
+    store3.addItem(2, 9.44, "Natrel Whole Milk, 32 fl oz");
```

Comment:
Extract store id, store name, address, item id, price, item name into const variables. And use the variables in the expected string below. Therefore you don't need to modify two places if you want to change the dataset in the future. Also it's more clearer to readers which items and stores are put into the expected string.

For example:

private static final int STORE_ID_1 = 1;
private static final String STORE_NAME_1 = "Walmart";

...

String expected = String.format("\"storeId\":%d,\"storeName\":%s", STORE_ID_1, STORE_NAME_1);

Reply:
I think because the responses are always changing right now due to updating schemas and varying data, my testing strategy was to print out the JSON response and validate it by thoroughly reading it and then just copying it into the test. This is a strategy I learned at a previous job when the test responses are too long to explicitly type out, and they are often changing.

What do you think?

Reply:
well right now you're hardcoding the variables (and not reading from the database), so when there's database changes this test will still fail? (e.g. if Horizon Milk disappears, this test won't pass). 

So more concretely, this test will look something like 
```
private static final String HORIZON_MILK = "Organic Whole Shelf-Stable Milk, 8 Oz., 12 Count";
store1.addItem("1", 11.98, HORIZON_MILK, "Horizon", "MILK"); // same for other variables

...
 String expected = String.format("[{\"lowestPotentialPrice\":22.28,\"totalItemsFound\":1,\"storeId\":1,\"items\":{\"MILK\":[{\"itemId\":\"1\",\"storeId\":1,\"itemPrice\":11.98,\"itemBrand\":\"Horizon\",\"itemName\":\"%s\"},{\"itemId\":\"2\",\"storeId\":1,\"itemPrice\":10.3,\"itemBrand\":\"Natrel\",\"itemName\":\"Whole Milk, 32 fl oz\"}]},\"typeToPrice\":{\"MILK\":10.3},\"storeAddress\":\"3255 Mission College Blvd\",\"storeName\":\"Walmart\"},", HORIZON_MILK);
...
```